sky/skylet/providers/ibm/vpc_provider.py

```diff
@@ -0,0 +1,646 @@
+        return target_namespace_id
+        
+    def _get_cloud_functions_actions(self, namespace_id):
+        """returns meta data on namespaces of ibm cloud functions within a specified region
+        :param offset - offset from the beginning of the list of results attained from the GET request,
+                        which may contain up to 200 namespaces per http response"""
+
+        res = requests.get(f"https://{self.namespace_region}.functions.cloud.ibm.com/api/v1/namespaces/{namespace_id}/actions?limit=200", headers=self.get_headers())
+        return json.loads(res.text)
+
+    def create_action(self, namespace_id):
+        logger.info(f"creating action on namespace: {namespace_id}")
+        # Define the function parameters
+        function_params = {
+            "exec": {
+                "kind": "python:3.9",
+                "code": self.function_code
+            },
+            "limits":{
+                "timeout":600000
+            }
+        }
+        res = requests.put(f"https://{self.namespace_region}.functions.cloud.ibm.com/api/v1/namespaces/{namespace_id}/actions/{self.action_name}?blocking=true&overwrite=true", headers=self.get_headers(), data=json.dumps(function_params))
+        if res.status_code!=200:
+            logger.error(res.text)        
+        return json.loads(res.text)
+
+    def delete_action(self, namespace_id):
+        """return the deleted function's metadata if it existed."""
+        logger.info(f"deleting action on namespace: {namespace_id}")
+        res = requests.delete(f"https://{self.namespace_region}.functions.cloud.ibm.com/api/v1/namespaces/{namespace_id}/actions/{self.action_name}?blocking=true", headers=self.get_headers())
+        if res.status_code!=200:
+            logger.warn(res.text)
+        return json.loads(res.text)
+
+    def invoke_action(self, namespace_id:str):
+        logger.info(f"invoking action on namespace: {namespace_id}")
+        payload = {'iam_api_key':ibm.get_api_key(),'resource_group_id':self.resource_group_id, 'vpc_id':self.vpc_id, 'region': self.vpc_region}
+        res = requests.post(f"https://{self.namespace_region}.functions.cloud.ibm.com/api/v1/namespaces/{namespace_id}/actions/{self.action_name}?blocking=true", headers=self.get_headers(), data=json.dumps(payload))
+        if res.status_code!=200:
+            logger.error(res.text)
+        return json.loads(res.text)
+
+    def delete_cluster(self):
+        """Deletes VPC with id==self.vpc_id.
+        1. creates a CloudFunctions namespace named ClusterCleaner.namespace_name if doesn't exists.
+        2. using idempotent function that deletes an action named ClusterCleaner.action_name if exists.
+        3. invokes the action to delete the VPC and all its resources."""
+        cf_namespace_id = self.create_or_fetch_namespace()
+        self.delete_action(cf_namespace_id)
+        self.create_action(cf_namespace_id)
+        self.invoke_action(cf_namespace_id)
```

Comment:
Out of curiosity, how do we ensure that the action runs __after__ the head node is down? I'm not familiar with IBM cloud functions.

Reply:
The action itself is entrusted with the job of taking down the the cluster's nodes, afterwards taking down the rest of the VPC's resources. no other workflow will try to take down the head node (or any other node). 
the following (documented) condition will make sure of that:
```
  # if the request is executed by the remote head like in `sky autostop --down`
  # , in contrast to `sky down`, use cloud functions to delete the VPC
  # with its associated resources. 
  hostname = socket.gethostname()  # returns the instance's (VSI) name
  if self._get_node_type(hostname) == NODE_KIND_HEAD:
      self.vpc_provider.remote_cluster_removal(vpc_id_to_delete, self.region)
```